README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
Quick start
-----------

Welcome to the `factordist` GitHub page!

This package measures the difference between factors. Most functions contained here work by tallying how often the levels of a factor equal one another.

``` r
library(devtools)
devtools::install_github("tpq/factordist")
library(factordist)
?factordist
```

The `factordist` function
-------------------------

The `factordist` function computes the difference between columns of a `data.frame`.

``` r
x <- simulate_factor_data(30, 6)
dis <- factordist(x, metric = "jaccard", as_dist = TRUE)
#> |------------(25%)----------(50%)----------(75%)----------|
```

If the columns do not contain factors already, the input is coerced to a factor.

``` r
x <- as.data.frame(matrix(1:25, 5, 5))
dis <- factordist(x, metric = "jaccard", as_dist = TRUE)
#> Alert: Coercing all non-factor input into factors.
#> |------------(25%)----------(50%)----------(75%)----------|
```

See `?factordist` for Details about available metrics. I have only implemented a few distance metrics so far, but welcome Feature Requests via <http://github.com/tpq/factordist/issues>.

The `cluster_model` function
----------------------------

The package also contains the `cluster_model` class and methods which are intended to make it easier to generalize clustering results to new data. Its design mirrors methods used for supervised learning. First, an initial function builds a model that relates the input data `x` to the labels `y`. Second, another function deploys the model to `newdata`, for which labels are unknown or hidden. When using the `predict` method, the labels for the new data get assigned procedurally using a nearest neighbor algorithm.

``` r
library(factordist)
x <- simulate_factor_data(30, 6)
y <- sample(c("A", "B"), replace = TRUE, size = 30)
newdata <- simulate_factor_data(10, 6)
m <- cluster_model(x, y, FUN = d_jaccard)
#> Alert: Coercing input y to string.
predict(m, newdata)
#> |------------(25%)----------(50%)----------(75%)----------|
#> Some samples have multiple nearest neighbors. Choosing randomly.
#>   1   2   3   4   5   6   7   8   9  10 
#> "A" "A" "B" "B" "B" "A" "B" "A" "A" "A"
```

By default, `cluster_model` will assume the metric is symmetric. Set the argument `asym = TRUE` if the metric is not symmetric, i.e., `dist(a, b) != dist(b, a)`. When `asym = TRUE`, the `predict` method will choose each neighbor based on the smaller of the two distances.


NEWS.md
## factordist 0.1.2
---------------------
* Add `as_consensus_cluster` function to suggest final labels from consensus matrix

## factordist 0.1.1
---------------------
* Set `strict = TRUE` as default `as_consensus_matrix` behavior
* Update `mode` to ignore NAs

## factordist 0.1.0
---------------------
* Add `as_consensus_matrix` function to repeat `match_labels` across matrix columns
* Clean up file structure and documentation
* Changes to streamline backend
    * Rename `ready_data` to `as_factor_frame`
    * New `resample` to avoid the `sample(N)` bug
    * New `which_max` and `which_min` to return sample max/min index
    * New `args_as_list` and `do_call` to simplify calling

## factordist 0.0.11
---------------------
* Add `match_labels` function to match cluster assignments
* Update `s_relAcc`
    * Fix denominator to ignore double NAs more precisely

## factordist 0.0.10
---------------------
* Add `s_relAcc` function to compute relative accuracy

## factordist 0.0.9
---------------------
* Add unit test for `sym` functions

## factordist 0.0.8
---------------------
* Fix performance bottleneck in `predict.cluster_model` method

## factordist 0.0.7
---------------------
* Add `circularize` function

## factordist 0.0.6
---------------------
* Update `cluster_model`
    * Pre-process `factordist` and `cluster_model` input the same way
* Update `predict.cluster_model`
    * Pre-process `factordist` and `predict.cluster_model` input the same way
    * Add progress bar
* General changes
    * Use `inherits` for class checks

## factordist 0.0.5
---------------------
* Update `factordist`
    * Coerce input into factors with the same levels
    * Alert user if input contains NA values
* Update `cluster_model`
    * When x and y are data.frames, check that row names match

## factordist 0.0.4
---------------------
* Update `predict.cluster_model`
    * Assign NA label when there are no nearest neighbors
    * Now handles asymmetric difference functions
    * Now handles character labels

## factordist 0.0.3
---------------------
* Add `cluster_model` class and methods

## factordist 0.0.2
---------------------
* Add `s_union`, a set overlap similarity metric
* Add `d_jaccard`, a set overlap difference metric
* Update `factordist` front-end for Jaccard

## factordist 0.0.1
---------------------
* Add `factordist` front-end function
* Add `papply` to standardize difference function deployment
* Add `d_adjRand`, a meta-clustering difference metric


R/9-global.R
#' Coerce Data Frame Into "Factor Frame"
#'
#' This backend function coerces any matrix or data.frame into a
#'  new data.frame where each column is a factor.
#'
#' @param data Any matrix or data.frame.
#' @return A data.frame where each column is a factor.
#' @export
as_factor_frame <- function(data){

  if(inherits(data, "matrix")) data <- data.frame(data, stringsAsFactors = FALSE)
  if(!inherits(data, "data.frame")) stop("Input must be a data.frame.")
  if(any(is.na(data))) message("Alert: NAs detected. NA handling depends on metric.")
  isfact <- sapply(data, is.factor)
  issamelevel <- sapply(data, function(x) identical(levels(x), levels(data[,1])))

  # Coerce input into factors that all have the same levels
  if(any(!isfact) | any(!issamelevel)){
    message("Alert: Coercing all factors to have same levels.")
    all_levels <- sort(unique(unlist(data)))
    for(i in 1:ncol(data)){
      data[,i] <- factor(data[,i], levels = all_levels)
    }
  }

  data
}

#' Print Progress Bar
#'
#' This backend function prints a progress bar.
#'
#' @param i The current iteration.
#' @param k Total iterations.
#' @param numTicks The result of \code{progress}.
#' @return The next \code{numTicks} argument.
#' @export
progress <- function(i, k, numTicks){

  if(i == 1) numTicks <- 0

  if(numTicks == 0) cat("|-")

  while(i > numTicks*(k/40)){

    cat("-")
    if(numTicks == 10) cat("(25%)")
    if(numTicks == 20) cat("(50%)")
    if(numTicks == 30) cat("(75%)")
    numTicks <- numTicks + 1
  }

  if(i == k) cat("-|\n")

  numTicks
}

#' A Simple(r) \code{sample} Function
#'
#' See \code{?sample}.
#'
#' @param x A vector.
#' @param ... Arguments to \code{sample}.
#' @return A vector.
#' @export
resample <- function(x, ...){

  x[sample.int(length(x), ...)]
}

#' Get Index of Maximum Value
#'
#' @param x A vector.
#' @return The index of the maximum of the vector. If there are
#'  multiple maximums, returns an index randomly.
#' @export
which_max <- function(x){

  resample(which(x == max(x)))[1]
}

#' Get Index of Minimum Value
#'
#' @param x A vector.
#' @return The index of the minimum of the vector. If there are
#'  multiple minimum, returns an index randomly.
#' @export
which_min <- function(x){

  resample(which(x == min(x)))[1]
}

#' Get Mode
#'
#' The mode will ignore NAs. If all elements are NA, the mode is NA.
#'
#' @param x A vector.
#' @return The mode of the vector. If there are multiple modes,
#'  returns a mode randomly.
#' @export
mode <- function(x){

  if(all(is.na(x))){

    x_mode <- NA

  }else{

    x <- x[!is.na(x)]
    counts <- table(x)
    x_mode <- names(which_max(counts))
  }

  x_mode
}

#' Coerce \code{...} to List
#'
#' @param ... Arguments.
#' @return A list of arguments.
#' @export
args_as_list <- function(...){

  as.list(eval(list(...)))
}

#' A Simple(r) \code{do.call} Function
#'
#' @param FUN A function to call.
#' @param ... Arguments.
#' @param .prior_arg_list More arguments. Used to pass an argument
#'  list collected from \code{\link{args_as_list}}. Unlike \code{...},
#'  this can be used to pass empty arguments, by setting
#'  \code{.prior_arg_list = list()}.
#' @return The result of \code{FUN}.
#' @export
do_call <- function(FUN, ..., .prior_arg_list = list()){

  args <- args_as_list(...)
  args <- append(args, .prior_arg_list)
  do.call(FUN, args)
}
